lib/api/api_auth_webauthn.go

```diff
@@ -0,0 +1,470 @@
+// Copyright (C) 2023 The Syncthing Authors.
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+package api
+
+import (
+	"bytes"
+	"encoding/base64"
+	"fmt"
+	"net/http"
+	"reflect"
+	"slices"
+	"time"
+
+	webauthnProtocol "github.com/go-webauthn/webauthn/protocol"
+	webauthnLib "github.com/go-webauthn/webauthn/webauthn"
+	"github.com/syncthing/syncthing/lib/config"
+	"github.com/syncthing/syncthing/lib/db"
+	"github.com/syncthing/syncthing/lib/events"
+	"github.com/syncthing/syncthing/lib/sliceutil"
+)
+
+func newWebauthnEngine(cfg config.Wrapper) (*webauthnLib.WebAuthn, error) {
+	guiCfg := cfg.GUI()
+
+	displayName := "Syncthing"
+	if dev, ok := cfg.Device(cfg.MyID()); ok && dev.Name != "" {
+		displayName = "Syncthing @ " + dev.Name
+	}
+
+	rpId := guiCfg.WebauthnRpId
+	if rpId == "" {
+		guiCfgStruct := reflect.TypeOf(guiCfg)
+		field, found := guiCfgStruct.FieldByName("WebauthnRpId")
+		if !found {
+			return nil, fmt.Errorf(`Field "WebauthnRpId" not found in struct GUIConfiguration`)
+		}
+		rpId = field.Tag.Get("default")
+		if rpId == "" {
+			return nil, fmt.Errorf(`Default tag not found on field "WebauthnRpId" in struct GUIConfiguration`)
+		}
```

Comment:
Oh, it does? I'm fairly sure I saw that and tried it, but that I couldn't get it to work for some reason. Might have been that the defaults only apply when unmarshaling a protobuf, but not when constructing struct literals in tests, or something like that. I'll take another look.

Reply:
Yeah, I added this because the defaults defined in protobuf don't apply to the config mocks used in tests. I didn't want to hard-code the same defaults in two places (in protobuf and here) as that invites them to go out of sync at some point. I preferred having the default in protobuf rather than in this constructor so that it assigns the default value when missing, so that the default is explicitly visible in the settings GUI, so that it's easier to notice if it's misconfigured than if empty values fall back to some invisible default.

Maybe the latter isn't as much of a concern since these are advanced settings, though. Code-wise, the easiest fix for this would be to move the defaults from the proto definition to the `newWebauthnEngine` constructor - simply falling back to hard-coded defaults when the config values are empty. That would make the "invisible defaults" issue I mentioned above, but maybe that's fine?

Reply:
I was wrong, it does nothing of the sort, it just adds the default value as a tag on the generated go struct. And then we have `structutil.SetDefault` to apply those, which we do when creating a fully new config. If you need defaults in tests, you can use that function on the mocked config (we should probably anyway do that in many places for more "realistic"/relevant testing). Would that work for the concrete problem you encounter?

Reply:
Ah, right. I tried this out and found that it doesn't quite work as a catch-all solution, because `structutil.SetDefault` unconditionally overwrites all properties with a default tag, not just those with a zero value. But I think I found a fair enough solution anyway, which is now included in [emlun#4](https://github.com/emlun/syncthing/pull/4/files/e908335fd765abe7cefe0bf2193b084239522ee2..fa16d27913ec4143e0fe627cf1917a06d31192cf).